ENH: Added public accessor registrar

TomAugspurger opened the pull request:
Adds new methods for registing custom accessors to pandas objects.

This will be helpful for implementing https://github.com/pandas-dev/pandas/issues/18767
outside of pandas.

If we accept this I'm not sure it belongs in the top-level namespace where I have it currently. What would a good home be for this be? `pd.api.extensions`?

Closes https://github.com/pandas-dev/pandas/issues/14781

TomAugspurger committed:
ENH: Added public accessor registrar

Adds new methods for registing custom accessors to pandas objects.

This will be helpful for implementing https://github.com/pandas-dev/pandas/issues/18767
outside of pandas.

Closes https://github.com/pandas-dev/pandas/issues/14781

pep8speaks commented:
Hello @TomAugspurger! Thanks for submitting the PR.





 - In the file [`pandas/errors/__init__.py`](https://github.com/pandas-dev/pandas/blob/6ae52d4e1a1da2b0b7db6cf7274901f44e98db83/pandas/errors/__init__.py), following are the PEP8 issues :

> [Line 71:68](https://github.com/pandas-dev/pandas/blob/6ae52d4e1a1da2b0b7db6cf7274901f44e98db83/pandas/errors/__init__.py#L71): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

TomAugspurger committed:
PEP8

gfyoung commented:
@TomAugspurger : A couple of points:

1) Re: your question, it depends on how much importance you put on the resource.  Personally, I'm fine with exposing it in the top-level namespace.  `xarray` does this, so one could argue that we should do the same for consistency.  Unless you're using this solely for internal stuff, then I don't see why not.

2) Could you explain the differences / changes you needed to make for this to port over to `pandas` ?  That would be very useful for understanding why you needed to add this to `pandas`.

gfyoung commented on doc/source/internals.rst:
```diff
@@ -89,6 +89,46 @@ not check (or care) whether the levels themselves are sorted. Fortunately, the
+      >>> ds.geo.center
+      (5.0, 10.0)
+      >>> ds.geo.plot()
+      # plots data on a map
```
To make this less abstract, you probably will need plots and some more concrete code to actually plot these points.

TomAugspurger replied:
I'm OK with this being abstract. The point is the accessor, not the functionality provided by it, if that makes sense.

jbrockmendel commented:
There might be some useful things to pull out of the [tests](https://github.com/pandas-dev/pandas/pull/17042/files#diff-6df9fe81c485ca477e339301a31cc53b) for the now-abandoned #17042.  I wrote a moderately fleshed out example for that.

shoyer commented on pandas/core/accessor.py:
```diff
@@ -129,3 +134,109 @@ def f(self, *args, **kwargs):
+
+
+# Ported with modifications from xarray
+# https://github.com/pydata/xarray/blob/master/xarray/core/extensions.py
```
Indeed, please do let me know exactly what you needed to change :)

gfyoung replied:
> Here's the relevant diff between xarray and pandas:

@TomAugspurger : For better or worse, that's a much less significant change than I expected.  Maybe should add a comment about that.

jreback commented:
this should not be in the main public namespace.

TomAugspurger commented:
Here's the relevant diff between xarray and pandas:

TomAugspurger committed:
Moved to extensions

TomAugspurger committed:
More docs

TomAugspurger committed:
Fix see also

jreback commented on pandas/core/accessor.py:
```diff
@@ -129,3 +134,135 @@ def f(self, *args, **kwargs):
+# https://github.com/pydata/xarray/blob/master/xarray/core/extensions.py
+
+
+class _CachedAccessor(object):
```
Is there a reason you are not using this for the internal accessors?

jreback commented on pandas/extensions/__init__.py:
```diff
@@ -0,0 +1,4 @@
+"""Public API for extending panadas objects."""
```
this is better, but still not a fan of extending the top-level namespaces (e.g. with extensions). api.extensions would be better

jreback commented on pandas/tests/api/test_api.py:
```diff
@@ -41,6 +41,13 @@ class TestPDApi(Base):
     # misc
     misc = ['IndexSlice', 'NaT']
 
+    # extension points
```
see above

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into accessor-decorator

jreback commented on pandas/tests/test_register_accessor.py:
```diff
@@ -0,0 +1,75 @@
+from pandas.errors import AccessorRegistrationWarning
+
+
+@contextlib.contextmanager
```
you can just use the pytest monkeypatch fixture for this

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into accessor-decorator

TomAugspurger committed:
DOC: Added whatsnew

TomAugspurger committed:
Move to api

TomAugspurger committed:
Update post review

TomAugspurger committed:
flake8

TomAugspurger committed:
Raise the underlying error instead of a RuntimeError

jreback commented:
> I have a slight preference to focusing this PR on just adding the register_*_accessor, but I can also refactor the internals to use it. Really though, these are independent. The only real change would be to pd.api.extensions.resgister_series_accessor("str")(StringMethods) instead of str = accessor.AccessorProperty(strings.StringMethods). The register* decorators are just for registering, not for implementing accessors themselves.

I disagree. The best use of this is to actually implement the internals. If it works, great, if not then you have found a bug / feature lacking.

TomAugspurger committed:
str validate

TomAugspurger committed:
DOC: Moved to developer

TomAugspurger committed:
REF: Use public registrars for accessors

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into accessor-decorator

TomAugspurger committed:
cleanup

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into accessor-decorator

TomAugspurger committed:
Implemented optional caching

TomAugspurger committed:
Document cache

TomAugspurger committed:
Tests passing

TomAugspurger committed:
Use for plot

TomAugspurger committed:
Fix autodoc

TomAugspurger committed:
Fix the class instantiation

TomAugspurger commented:
Refactored to have our accessors use it.

1.  All the accessors are registered in `pandas.core.register_accessors` after the classes have been initialized
2. I ran into a subtle issue around caching. The accessor from xarray is cached such that `x.foo` is the same object everytime. With our old ``AccessorProperty``, we recreated the accessor every time. This caused some issues when the object itself was mutable, but the accessor cached some state. Notably, this would occur:

```python
In [9]: s = pd.Series(pd.TimedeltaIndex(['5S', '10S']))

In [10]: s.dt.values.isna()
Out[10]: array([False, False], dtype=bool)

In [11]: s[0] = float('nan')

# isna has been cached on the accessor's .values, but not invalidated
# when s was mutated
In [12]: s.dt.values.isna()
Out[12]: array([ False, False], dtype=bool)
```

To work around this, I added a `cache` keyword argument (@shoyer would xarray be interested in this keyword?)

3. Moved the new docs to "developer.rst" instead of internals, since it's for "downstream applications" (we should probably move the subclassing docs there)
4. Removed the now unused AccessorProperty class

shoyer commented:
> I ran into a subtle issue around caching. The accessor from xarray is cached such that x.foo is the same object everytime. With our old AccessorProperty, we recreated the accessor every time. This caused some issues when the object itself was mutable, but the accessor cached some state. 
>
> To work around this, I added a cache keyword argument (@shoyer would xarray be interested in this keyword?)

This seems like an inadvertent detail of the implementation of `Series.dt.values` (why do we even have that property?). The datetime accessor creates a DatetimeIndex internally for method/properties. `DatetimeIndex.values` is fixed, because pandas.Index objects are immutable.

So the other fix would be to change how the `.dt` accessor implemented so that `DatetimeIndex` objects are created on demand when a method/property is called, instead of being created once and saved on the acccessor.

I don't think we need a `cache=False` option for xarray since nobody has written any accessors relying on this behavior.

jbrockmendel commented:
Making it easy to extend functionality is great, but the new commit that changes where the current str/cat/dt attributes are defined I'm very much -1 on.  Inheritance notwithstanding, the class definition should be WYSIWYG.  Newbies shouldn't have to [track down](https://github.com/pandas-dev/pandas/pull/17174/files#diff-de56e43aab6f9b0909c90246728a3662L207) where e.g. `DataFrame.pivot_table` is defined.

TomAugspurger commented:
> This seems like an inadvertent detail of the implementation of Series.dt.values (why do we even have that property?).

`.values` isn't actually in `dir(s.dt)`. Should probably be `._values` but that's how it was before. Regardless, I'll see how difficult it is to create them on demand.

> the class definition should be WYSIWYG

That's my preference as well, we just won't be able to / have to use the `register_*_accessor` functions, since those rely on the classes being defined.

TomAugspurger committed:
Refactor again.

1. Removed optional caching
2. Refactored `Properties` to create the indexes it uses on demand
3. Moved accessor definitions to classes for clarity

TomAugspurger committed:
Fix API files

TomAugspurger committed:
Remove stale comment

TomAugspurger commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -119,6 +119,56 @@ Current Behavior
     s.rank(na_option='top')
 
+
+Extending Pandas Objects with New Accessors
```
Thoughts on removing this section in the whatsnew? This may be too niche for most users to care about.

TomAugspurger committed:
Tests pass

TomAugspurger committed:
DOC: some cleanup

TomAugspurger committed:
No need to assign doc

jreback commented on pandas/core/accessor.py:
```diff
@@ -129,3 +102,136 @@ def f(self, *args, **kwargs):
+# 1. We don't need to catch and re-raise AttributeErrors as RuntimeErrors
+
+
+class _CachedAccssor(object):
```
name this: ``CachedAccessor`` we use this internally so private is not appropriate

jreback commented on pandas/core/accessor.py:
```diff
@@ -129,3 +102,136 @@ def f(self, *args, **kwargs):
+        # NDFrame
+        object.__setattr__(obj, self._name, accessor_obj)
+        return accessor_obj
+
```
it might make sense to have a few methods here which are AbstractMethodError to guide the api, eg. ``validate``

TomAugspurger replied:
I'd prefer to leave that open to the person implementing the accessor. Perhaps for some accessors, there's no need for validation.

jreback replied:
this is too general we need to be opionated, if there is no validation then that should be explicit.

TomAugspurger replied:
Not sure what you mean be "too general". The purpose of this is to give an officially blessed way to do "DataFrame.foo = MyAccessor"; what you do past that point is up to the library. It's not like `CachedAccessor` ever has a chance to see the data so that it can validate things. 

The only restriction on the accessor is that its init method be `__init__(self, pandas_obj)`. I'll document that, but I'm not sure what else there is to be done.

jreback commented on pandas/core/categorical.py:
```diff
@@ -2161,12 +2165,19 @@ class CategoricalAccessor(PandasDelegate, PandasObject, NoNewAttributesMixin):
-        self.index = index
-        self.name = name
+    def __init__(self, data):
+        self._validate(data)
```
see my commet above, these can be non-private I think as these are not user exposed

jreback commented on pandas/core/indexes/accessors.py:
```diff
@@ -20,81 +21,44 @@
-    data : Series
-    copy : boolean, default False
-           copy the input data
+    def _get_values(self):
```
generally would like these to be non-private

TomAugspurger replied:
I think it's best for methods and attributes on the accessor to be private. Imagine if we wanted to delegate a method like `.get_values`, so `foo.bar.get_values`. This would break the accessor, since its `get_values` would be broken.

jreback commented on pandas/core/indexes/accessors.py:
```diff
@@ -245,15 +210,34 @@ class PeriodProperties(Properties):
 
-    @classmethod
-    def _make_accessor(cls, data):
+    def __new__(cls, data):
```
can we be consistent for validation, IOW either do it in ``__new__`` or in ``__init__``?

TomAugspurger replied:
I had to put this in `__new__` since we don't instantiate a `CombinedDatetimelikeProperties`, just one of its parents. I *think* this is unavoidable, but it would be nice to make it more consistent with the others. I'll see what I can do.

TomAugspurger replied:
One option is to write a `Properties._validate` do most of the work here. But it feels weird to have a base class decide which child class should be instantiated. I think having it in `__new__` (with a comment for why it's there) is clearest.

jreback reviewed (changes requested):
looks pretty good. some api consistency questions.

TomAugspurger committed:
Rename, shared docs

TomAugspurger committed:
Doc __new__

TomAugspurger commented:
@jreback @jbrockmendel thoughts on the refactored version? Anything else you'd like to see changed?

jbrockmendel commented:
> thoughts on the refactored version?

Completely addresses my WYSIWYG complaint, thanks.

TomAugspurger commented:
>  this is too general we need to be opionated, if there is no validation then that should be explicit.

@jreback could you clarify your thoughts here? I've documented that the only requirement on the user-defined accessor is that it's `__init__` will be called with `pandas_obj`. I don't think there's anything the `_CachedAccessor` can / should do w.r.t. validation. With an accessor like `.plot` there's nothing to be done, so it doesn't make sense for accessors to implement a method that does nothing.

jreback commented on pandas/errors/__init__.py:
```diff
@@ -65,3 +65,7 @@ class MergeError(ValueError):
     """
+
+
+class AccessorRegistrationWarning(Warning):
```
do we really need a special warning here?

TomAugspurger replied:
Switched to a UserWarning, which matches what we do with `df.x = [1, 2, 3]`.

jreback reviewed (changes requested):
ok, looks fine. one question on the warning.

TomAugspurger committed:
Use UserWarning

TomAugspurger committed:
Update test

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into accessor-decorator

TomAugspurger commented:
FYI, I haven't thought through whether `PeriodArray` / `IntervalArray` should be part of the public API in the same way that `Categorical` is. One the one hand, you'll get them when you do `pd.Series(pd.period_range("2000", periods=10))`, so they should probably be documented, etc. On the other hand, it's just one more thing to worry about. Is it useful to generate a `pandas.PeriodArray(...)`  on it's own?

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into accessor-decorator
